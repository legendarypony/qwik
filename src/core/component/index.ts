/**
 * @license
 * Copyright Builder.io, Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/BuilderIO/qwik/blob/main/LICENSE
 */

/**
 * @fileoverview
 *
 * Explicitly list public exported symbols to be used by the developer.
 */

export { Component, ComponentConstructor, ComponentPropsOf, ComponentStateOf } from './component';
