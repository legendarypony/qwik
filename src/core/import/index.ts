/**
 * @license
 * Copyright Builder.io, Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/BuilderIO/qwik/blob/main/LICENSE
 */

export { qImport } from './qImport.js';
export { QRL } from './qrl.js';
