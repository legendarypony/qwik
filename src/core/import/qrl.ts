/**
 * @license
 * Copyright Builder.io, Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/BuilderIO/qwik/blob/main/LICENSE
 */

import { EMPTY_ARRAY } from '../util/flyweight';
import type { QRL } from './qrl.public';
import { isQrl, QRLInternal } from './qrl-class';
import { assertEqual } from '../assert/assert';
import type { CorePlatform } from '../index';
import { getDocument } from '../util/dom';
import { logError } from '../util/log';

let runtimeSymbolId = 0;
const RUNTIME_QRL = '/runtimeQRL';

// https://regexr.com/68v72
const EXTRACT_IMPORT_PATH = /\(\s*(['"])([^\1]+)\1\s*\)/;

// https://regexr.com/690ds
const EXTRACT_SELF_IMPORT = /Promise\s*\.\s*resolve/;

// https://regexr.com/6a83h
const EXTRACT_FILE_NAME = /[\\/(]([\w\d.\-_]+\.(js|ts)x?):/;

export function toInternalQRL<T>(qrl: QRL<T>): QRLInternal<T> {
  assertEqual(isQrl(qrl), true);
  return qrl as QRLInternal<T>;
}

export function staticQrl<T = any>(
  chunkOrFn: string | (() => Promise<any>),
  symbol: string,
  lexicalScopeCapture: any[] = EMPTY_ARRAY
): QRL<T> {
  let chunk: string;
  let symbolFn: null | (() => Promise<Record<string, any>>) = null;
  if (typeof chunkOrFn === 'string') {
    chunk = chunkOrFn;
  } else if (typeof chunkOrFn === 'function') {
    symbolFn = chunkOrFn;
    let match: RegExpMatchArray | null;
    const srcCode = String(chunkOrFn);
    if ((match = srcCode.match(EXTRACT_IMPORT_PATH)) && match[2]) {
      chunk = match[2];
    } else if ((match = srcCode.match(EXTRACT_SELF_IMPORT))) {
      const ref = 'QWIK-SELF';
      const frames = new Error(ref).stack!.split('\n');
      const start = frames.findIndex((f) => f.includes(ref));
      const frame = frames[start + 2];
      match = frame.match(EXTRACT_FILE_NAME);
      if (!match) {
        chunk = 'main';
      } else {
        chunk = match[1];
      }
    } else {
      throw new Error('Q-ERROR: Dynamic import not found: ' + srcCode);
    }
  } else {
    throw new Error('Q-ERROR: Unknown type argument: ' + chunkOrFn);
  }
  return new QRLInternal<T>(chunk, symbol, null, symbolFn, null, lexicalScopeCapture, null);
}

export function runtimeQrl<T>(symbol: T, lexicalScopeCapture: any[] = EMPTY_ARRAY): QRLInternal<T> {
  return new QRLInternal<T>(
    RUNTIME_QRL,
    's' + runtimeSymbolId++,
    symbol,
    null,
    null,
    lexicalScopeCapture,
    null
  );
}

export function stringifyQRL(qrl: QRL, platform?: CorePlatform, element?: Element) {
  const qrl_ = toInternalQRL(qrl);
  const symbol = qrl_.symbol;
  const chunk = platform ? platform.chunkForSymbol(symbol) ?? qrl_.chunk : qrl_.chunk;

  const parts: string[] = [chunk];
  if (symbol && symbol !== 'default') {
    parts.push('#', symbol);
  }
  const guard = qrl_.guard;
  guard?.forEach((value, key) =>
    parts.push('|', key, value && value.length ? '.' + value.join('.') : '')
  );
  const capture = qrl_.capture;
  if (capture && capture.length > 0) {
    parts.push(JSON.stringify(capture));
  }
  const qrlString = parts.join('');
  if (qrl_.chunk === RUNTIME_QRL && element) {
    const qrls: Set<QRL> = (element as any).__qrls__ || ((element as any).__qrls__ = new Set());
    qrls.add(qrl);
  }
  return qrlString;
}

export function qrlToUrl(element: Element, qrl: QRL): URL {
  return new URL(stringifyQRL(qrl), getDocument(element).baseURI);
}

/**
 * `./chunk#symbol|symbol.propA.propB|[captures]
 */
export function parseQRL(qrl: string, element?: Element): QRLInternal {
  if (element) {
    const qrls: QRLInternal[] | undefined = (element as any).__qrls__;
    if (qrls) {
      for (const runtimeQrl of qrls) {
        if (stringifyQRL(runtimeQrl) == qrl) {
          return runtimeQrl;
        }
      }
    }
  }
  const endIdx = qrl.length;
  const hashIdx = indexOf(qrl, 0, '#');
  const guardIdx = indexOf(qrl, hashIdx, '|');
  const captureIdx = indexOf(qrl, guardIdx, '[');

  const chunkEndIdx = Math.min(hashIdx, guardIdx, captureIdx);
  const chunk = qrl.substring(0, chunkEndIdx);

  const symbolStartIdx = hashIdx == endIdx ? hashIdx : hashIdx + 1;
  const symbolEndIdx = Math.min(guardIdx, captureIdx);
  const symbol =
    symbolStartIdx == symbolEndIdx ? 'default' : qrl.substring(symbolStartIdx, symbolEndIdx);

  const guardStartIdx = guardIdx;
  const guardEndIdx = captureIdx;
  const guard =
    guardStartIdx < guardEndIdx ? parseGuard(qrl.substring(guardStartIdx, guardEndIdx)) : null;

  const captureStartIdx = captureIdx;
  const captureEndIdx = endIdx;
  const capture =
    captureStartIdx === captureEndIdx
      ? EMPTY_ARRAY
      : JSONparse(qrl.substring(captureStartIdx, captureEndIdx));

  if (chunk === RUNTIME_QRL) {
    logError(`Q-ERROR: '${qrl}' is runtime but no instance found on element.`);
  }
  return new QRLInternal(chunk, symbol, null, null, capture, null, guard);
}

function JSONparse(json: string): any {
  try {
    return JSON.parse(json);
  } catch (e) {
    logError('JSON:', json);
    throw e;
  }
}

function parseGuard(text: string): null | Map<string, string[]> {
  let map: null | Map<string, string[]> = null;
  if (text) {
    text.split('|').forEach((obj) => {
      if (obj) {
        const parts = obj.split('.');
        const id = parts.shift()!;
        if (!map) map = new Map<string, string[]>();
        map.set(id, parts);
      }
    });
  }
  return map;
}
function indexOf(text: string, startIdx: number, char: string) {
  const endIdx = text.length;
  const charIdx = text.indexOf(char, startIdx == endIdx ? 0 : startIdx);
  return charIdx == -1 ? endIdx : charIdx;
}

export function toQrlOrError<T>(symbolOrQrl: T | QRL<T>): QRLInternal<T> {
  if (!isQrl(symbolOrQrl)) {
    if (typeof symbolOrQrl == 'function' || typeof symbolOrQrl == 'string') {
      symbolOrQrl = runtimeQrl(symbolOrQrl);
    } else {
      // TODO(misko): centralize
      throw new Error(`Q-ERROR Only 'function's and 'string's are supported.`);
    }
  }
  return symbolOrQrl as QRLInternal<T>;
}
