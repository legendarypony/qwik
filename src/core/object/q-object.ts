import { assertEqual } from '../assert/assert';
import { QError, qError } from '../error/error';
import { isQrl } from '../import/qrl-class';
import { notifyRender } from '../render/notify-render';
import { tryGetInvokeContext } from '../use/use-core';
import { isElement } from '../util/element';
import { logWarn } from '../util/log';
import { qDev, qTest } from '../util/qdev';
import { debugStringify } from '../util/stringify';
import { runWatch, WatchDescriptor } from '../watch/watch.public';

export type ObjToProxyMap = WeakMap<any, any>;
export type QObject<T extends {}> = T & { __brand__: 'QObject' };

const ProxyMapSymbol = Symbol('ProxyMapSymbol');

export function getProxyMap(doc: Document): ObjToProxyMap {
  let map = (doc as any)[ProxyMapSymbol];
  if (!map) {
    map = (doc as any)[ProxyMapSymbol] = new WeakMap();
  }
  return map;
}

export function qObject<T extends Object>(obj: T, proxyMap: ObjToProxyMap): T {
  assertEqual(unwrapProxy(obj), obj, 'Unexpected proxy at this location');
  if (obj == null || typeof obj !== 'object') {
    // TODO(misko): centralize
    throw new Error(
      `Q-ERROR: Only objects can be wrapped in 'QObject', got ` + debugStringify(obj)
    );
  }
  if (obj.constructor !== Object) {
    throw new Error(
      `Q-ERROR: Only objects literals can be wrapped in 'QObject', got ` + debugStringify(obj)
    );
  }
  return readWriteProxy(obj as any as QObject<T>, proxyMap);
}

export function _restoreQObject<T>(obj: T, map: ObjToProxyMap, subs: Map<Element, Set<string>>): T {
  return readWriteProxy(obj as any as QObject<T>, map, subs);
}

/**
 * Creates a proxy which notifies of any writes.
 */
export function readWriteProxy<T extends object>(
  target: T,
  proxyMap: ObjToProxyMap,
  subs?: Map<Element, Set<string>>
): T {
  if (!target || typeof target !== 'object') return target;
  let proxy = proxyMap.get(target);
  if (proxy) return proxy;

  proxy = new Proxy(target, new ReadWriteProxyHandler(proxyMap, subs)) as any as T;
  proxyMap.set(target, proxy);
  return proxy;
}

export const QOjectTargetSymbol = ':target:';
export const QOjectSubsSymbol = ':subs:';
export const QOjectOriginalProxy = ':proxy:';
export const SetSubscriber = Symbol('SetSubscriber');

export function unwrapProxy<T>(proxy: T): T {
  if (proxy && typeof proxy == 'object') {
    const value = (proxy as any)[QOjectTargetSymbol];
    if (value) return value;
  }
  return proxy;
}

export function wrap<T>(value: T, proxyMap: ObjToProxyMap): T {
  if (value && typeof value === 'object') {
    if (isQrl(value)) {
      return value;
    }
    const nakedValue = unwrapProxy(value);
    if (nakedValue !== value) {
      // already a proxy return;
      return value;
    }
    if (qDev) {
      verifySerializable<T>(value);
    }
    const proxy = proxyMap.get(value);
    return proxy ? proxy : readWriteProxy(value as any, proxyMap);
  } else {
    return value;
  }
}

type TargetType = Record<string | symbol, any>;

class ReadWriteProxyHandler implements ProxyHandler<TargetType> {
  private subscriber?: Element;
  constructor(
    private proxyMap: ObjToProxyMap,
    private subs = new Map<Element | WatchDescriptor, Set<string>>()
  ) {}

  getSub(el: Element) {
    let sub = this.subs.get(el);
    if (!sub) {
      this.subs.set(el, (sub = new Set()));
    }
    return sub;
  }

  get(target: TargetType, prop: string | symbol): any {
    let subscriber = this.subscriber;
    this.subscriber = undefined;
    if (prop === QOjectTargetSymbol) return target;
    if (prop === QOjectSubsSymbol) return this.subs;
    if (prop === QOjectOriginalProxy) return this.proxyMap.get(target);
    const value = target[prop];
    if (typeof prop === 'symbol') {
      return value;
    }
    if (!subscriber) {
      const invokeCtx = tryGetInvokeContext();
      if (qDev && !invokeCtx && !qTest) {
        logWarn(`State assigned outside invocation context. Getting prop "${prop}" of:`, target);
      }
      if (invokeCtx && invokeCtx.subscriptions && invokeCtx.hostElement) {
        subscriber = invokeCtx.hostElement;
      }
    }
    if (subscriber) {
      const isArray = Array.isArray(target);
      const sub = this.getSub(subscriber);
      if (!isArray) {
        sub.add(prop);
      }
    }
    return wrap(value, this.proxyMap);
  }

  set(target: TargetType, prop: string | symbol, newValue: any): boolean {
    if (typeof prop === 'symbol') {
      if (prop === SetSubscriber) {
        this.subscriber = newValue;
      } else {
        target[prop] = newValue;
      }
      return true;
    }
    const unwrappedNewValue = unwrapProxy(newValue);
    verifySerializable(unwrappedNewValue);
    const isArray = Array.isArray(target);
    if (isArray) {
      target[prop as any] = unwrappedNewValue;
      this.subs.forEach((_, sub) => notifyChange(sub));
      return true;
    }
    const oldValue = target[prop];
    if (oldValue !== unwrappedNewValue) {
      target[prop] = unwrappedNewValue;
      this.subs.forEach((propSets, sub) => {
        if (propSets.has(prop)) {
          notifyChange(sub);
        }
      });
    }
    return true;
  }

  has(target: TargetType, property: string | symbol) {
    if (property === QOjectTargetSymbol) return true;
    if (property === QOjectSubsSymbol) return true;

    return Object.prototype.hasOwnProperty.call(target, property);
  }

  ownKeys(target: TargetType): ArrayLike<string | symbol> {
    return Object.getOwnPropertyNames(target);
  }
}

export function notifyChange(subscriber: Element | WatchDescriptor) {
  if (isElement(subscriber)) {
    notifyRender(subscriber);
  } else {
    runWatch(subscriber as WatchDescriptor);
  }
}

function verifySerializable<T>(value: T) {
  if (shouldSerialize(value) && typeof value == 'object' && value !== null) {
    if (Array.isArray(value)) return;
    if (isQrl(value)) return;
    if (Object.getPrototypeOf(value) !== Object.prototype) {
      throw qError(QError.TODO, 'Only primitive and object literals can be serialized.');
    }
  }
}

const NOSERIALIZE = Symbol('NoSerialize');

export function shouldSerialize(obj: any): boolean {
  if (obj !== null && (typeof obj == 'object' || typeof obj === 'function')) {
    const noSerialize = (obj as any)[NOSERIALIZE] === true;
    return !noSerialize;
  }
  return true;
}

/**
 * @alpha
 */
export type NoSerialize<T> = (T & { [NOSERIALIZE]: true }) | undefined;

/**
 * @alpha
 */
export function noSerialize<T extends {}>(input: T): NoSerialize<T> {
  (input as any)[NOSERIALIZE] = true;
  return input as any;
}
