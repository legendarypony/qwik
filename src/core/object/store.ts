import { assertDefined } from '../assert/assert';
import { getContext } from '../props/props';
import { isElement } from '../util/element';
import {
  ELEMENT_ID,
  ELEMENT_ID_PREFIX,
  ELEMENT_ID_SELECTOR,
  QHostAttr,
  QObjAttr,
  QObjSelector,
} from '../util/markers';
import { qDev } from '../util/qdev';
import { QOjectSubsSymbol, QOjectTargetSymbol, _restoreQObject } from './q-object';

export interface Store {
  doc: Document;
  objs: Record<string, any>;
}

export const JSON_OBJ_PREFIX = '\u0010';

export function QStore_hydrate(doc: Document) {
  const script = doc.querySelector('script[type="qwik/json"]');
  (doc as any).qDehydrate = () => QStore_dehydrate(doc);
  if (script) {
    script.parentElement!.removeChild(script);
    const meta = JSON.parse(script.textContent || '{}') as any;
    const elements = new Map<string, Element>();
    doc.querySelectorAll(ELEMENT_ID_SELECTOR).forEach((el) => {
      const id = el.getAttribute(ELEMENT_ID)!;
      elements.set(ELEMENT_ID_PREFIX + id, el);
    });
    for (const obj of meta.objs) {
      reviveNestedQObjects(obj, meta.objs);
    }
    reviveQObjects(meta.objs, meta.subs, elements);

    doc.querySelectorAll(QObjSelector).forEach((el) => {
      const qobj = el.getAttribute(QObjAttr);
      const host = el.getAttribute(QHostAttr);
      const ctx = getContext(el);
      qobj!.split(' ').forEach((part) => {
        const obj = part[0] === ELEMENT_ID_PREFIX ? elements.get(part) : meta.objs[strToInt(part)];
        assertDefined(obj);
        ctx.refMap.add(obj);
      });
      if (host) {
        const [props, events] = host.split(' ').map(strToInt);
        assertDefined(props);
        assertDefined(events);
        ctx.props = ctx.refMap.get(props);
        ctx.events = ctx.refMap.get(events);
      }
    });
  }
}

/**
 * Serialize the current state of the application into DOM
 *
 * @param doc
 */
export function QStore_dehydrate(doc: Document) {
  const objSet = new Set<any>();

  // Element to index
  const elementToIndex = new Map<Element, string>();
  function getElementID(el: Element) {
    let id = elementToIndex.get(el);
    if (id === undefined) {
      id = intToStr(elementToIndex.size);
      el.setAttribute(ELEMENT_ID, id);
      id = ELEMENT_ID_PREFIX + id;
      elementToIndex.set(el, id);
    }
    return id;
  }
  // Find all Elements which have qObjects attached to them
  const elements = doc.querySelectorAll(QObjSelector);
  elements.forEach((node) => {
    const props = getContext(node);
    const qMap = props.refMap;
    qMap.array.forEach((v) => {
      collectQObjects(v, objSet);
    });
  });

  // Convert objSet to array
  const objArray = Array.from(objSet);
  objArray.sort((a, b) => {
    const isProxyA = a[QOjectTargetSymbol] !== undefined ? 0 : 1;
    const isProxyB = b[QOjectTargetSymbol] !== undefined ? 0 : 1;
    return isProxyA - isProxyB;
  });

  const objs = objArray.map((a) => {
    return a[QOjectTargetSymbol] ?? a;
  });

  const subs = objArray
    .map((a) => {
      const subs = a[QOjectSubsSymbol] as Map<Element, Set<string>>;
      if (subs) {
        return Object.fromEntries(
          Array.from(subs.entries()).map(([el, set]) => {
            if (el.isConnected) {
              const id = getElementID(el);
              return [id, Array.from(set)];
            } else {
              return [undefined, undefined];
            }
          })
        );
      } else {
        return null;
      }
    })
    .filter((a) => !!a);

  const objToId = new Map<any, number>();
  let count = 0;
  for (const obj of objs) {
    objToId.set(obj, count);
    count++;
  }

  const convert = (value: any) => {
    if (value && typeof value === 'object') {
      value = value[QOjectTargetSymbol] ?? value;
    }
    const idx = objToId.get(value);
    if (idx !== undefined) {
      return intToStr(idx);
    }
    return elementToIndex.get(value) ?? value;
  };

  const convertedObjs = objs.map((obj) => {
    if (Array.isArray(obj)) {
      return obj.map(convert);
    } else if (typeof obj === 'object') {
      const output: Record<string, any> = {};
      Object.entries(obj).forEach(([key, value]) => {
        output[key] = convert(value);
      });
      return output;
    }
    return obj;
  });

  const data = {
    objs: convertedObjs,
    subs,
  };

  // Write back to the dom
  elements.forEach((node) => {
    const ctx = getContext(node);
    const props = ctx.props;
    const events = ctx.events;
    const attribute = ctx.refMap.array
      .map((obj) => {
        if (isElement(obj)) {
          return getElementID(obj);
        }

        const idx =
          typeof obj === 'object' ? objToId.get(obj[QOjectTargetSymbol] ?? obj) : objToId.get(obj);

        assertDefined(idx);
        return intToStr(idx!);
      })
      .join(' ');
    node.setAttribute(QObjAttr, attribute);

    if (props) {
      const objs = [props];
      if (events) {
        objs.push(events);
      }
      node.setAttribute(QHostAttr, objs.map((obj) => ctx.refMap.indexOf(obj)).join(' '));
    }
  });

  // Serialize
  const script = doc.createElement('script');
  script.setAttribute('type', 'qwik/json');
  script.textContent = JSON.stringify(data, undefined, qDev ? '  ' : undefined);

  doc.body.appendChild(script);
}

function reviveQObjects(objs: object[], subs: any[], elementMap: Map<string, Element>) {
  for (let i = 0; i < objs.length; i++) {
    const sub = subs[i];
    if (sub) {
      const value = objs[i];
      const converted = new Map(
        Object.entries(sub).map((entry) => {
          const el = elementMap.get(entry[0])!;
          assertDefined(el);
          const set = new Set(entry[1] as any) as Set<string>;
          return [el, set];
        })
      );
      objs[i] = _restoreQObject(value, converted);
    }
  }
}

function reviveNestedQObjects(obj: any, map: object[]) {
  if (obj && typeof obj == 'object') {
    if (Array.isArray(obj)) {
      for (let i = 0; i < obj.length; i++) {
        const value = obj[i];
        if (typeof value == 'string') {
          obj[i] = map[strToInt(value)];
        } else {
          reviveNestedQObjects(value, map);
        }
      }
    } else {
      for (const key in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, key)) {
          const value = obj[key];
          if (typeof value == 'string') {
            obj[key] = map[strToInt(value)];
          } else {
            reviveNestedQObjects(value, map);
          }
        }
      }
    }
  }
}

function collectQObjects(obj: any, seen: Set<any>) {
  if (obj != null) {
    if (isElement(obj)) {
      return;
    }
    if (typeof obj === 'boolean') {
      return;
    }

    if (seen.has(obj)) return;
    seen.add(obj);

    if (typeof obj === 'object') {
      if (Array.isArray(obj)) {
        for (let i = 0; i < obj.length; i++) {
          collectQObjects(obj[i], seen);
        }
      } else {
        for (const key in obj) {
          if (Object.prototype.hasOwnProperty.call(obj, key)) {
            const value = obj[key];
            collectQObjects(value, seen);
          }
        }
      }
    }
  }
}

export const intToStr = (nu: number) => {
  return nu.toString(36);
};

export const strToInt = (nu: string) => {
  return parseInt(nu, 36);
};
