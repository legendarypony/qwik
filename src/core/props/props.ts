import { QError, qError } from '../error/error';
import { getProxyMap, readWriteProxy } from '../object/q-object';
import { QStore_hydrate } from '../object/store';
import type { RenderContext } from '../render/cursor';
import { getDocument } from '../util/dom';
import { newQObjectMap, QObjectMap } from './props-obj-map';
import { qPropWriteQRL, qPropReadQRL } from './props-on';

Error.stackTraceLimit = 9999;

// TODO(misko): For better debugger experience the getProps should never store Proxy, always naked objects to make it easier to traverse in the debugger.

const Q_IS_HYDRATED = '__isHydrated__';
export const Q_CTX = '__ctx__';

export function hydrateIfNeeded(doc: Document): void {
  const isHydrated = (doc as any)[Q_IS_HYDRATED];
  if (!isHydrated) {
    (doc as any)[Q_IS_HYDRATED] = true;
    QStore_hydrate(doc);
  }
}

export interface QContextEvents {
  [eventName: string]: string | undefined;
}

export interface QContext {
  cache: Map<string, any>;
  refMap: QObjectMap;
  element: Element;
  dirty: boolean;
  props: Record<string, any> | undefined;
  events: QContextEvents | undefined;
}

export function getContext(element: Element): QContext {
  let ctx: QContext = (element as any)[Q_CTX];
  if (!ctx) {
    const cache = new Map();
    (element as any)[Q_CTX] = ctx = {
      element,
      cache,
      refMap: newQObjectMap(element),
      dirty: false,
      props: undefined,
      events: undefined,
    };
  }
  return ctx;
}

export function setEvent(rctx: RenderContext, ctx: QContext, prop: string, value: any) {
  qPropWriteQRL(rctx, ctx, prop, value);
}

export function getEvent(ctx: QContext, prop: string): any {
  return qPropReadQRL(ctx, prop);
}

export function getProps(ctx: QContext) {
  if (!ctx.props) {
    ctx.props = readWriteProxy({}, getProxyMap(getDocument(ctx.element)));
    ctx.refMap.add(ctx.props);
  }
  return ctx.props!;
}

export function getEvents(ctx: QContext): QContextEvents {
  let events = ctx.events;
  if (!events) {
    events = ctx.events = {};
    ctx.refMap.add(ctx.events);
  }
  return events;
}

/**
 * Turn an `Array` or object literal into a `class` or `style`
 *
 * @param obj `string`, `Array` or object literal
 * @param isClass `true` if expecting `class` output
 * @returns `string`
 */
export function stringifyClassOrStyle(obj: any, isClass: boolean): string {
  if (obj == null) return '';
  if (typeof obj == 'object') {
    let text = '';
    let sep = '';
    if (Array.isArray(obj)) {
      if (!isClass) {
        throw qError(QError.Render_unsupportedFormat_obj_attr, obj, 'style');
      }
      for (let i = 0; i < obj.length; i++) {
        text += sep + obj[i];
        sep = ' ';
      }
    } else {
      for (const key in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, key)) {
          const value = obj[key];
          text += isClass ? (value ? sep + key : '') : sep + key + ':' + value;
          sep = isClass ? ' ' : ';';
        }
      }
    }
    return text;
  }
  return String(obj);
}
