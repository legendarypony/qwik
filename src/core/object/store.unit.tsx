import { createDocument } from '../../testing/document';
import { useStore } from '../use/use-store.public';
import { getContext, getProps } from '../props/props';
import type { Props } from '../props/props.public';
import { newInvokeContext, useInvoke } from '../use/use-core';
import { render } from '../render/render.public';
import { getQwikJSON } from './store';
import { useLexicalScope } from '../use/use-lexical-scope.public';
import { component$ } from '../component/component.public';
import { noSerialize } from './q-object';
import { $ } from '../import/qrl.public';
import { logDebug } from '../util/log';
import { runtimeQrl } from '../import/qrl';
import { pauseContainer } from '../object/store.public';

describe('store', () => {
  let document: Document;
  let div: HTMLElement;
  let qDiv: Props;

  beforeEach(() => {
    document = createDocument();
    div = document.createElement('div');
    document.body.appendChild(div);
    qDiv = getProps(getContext(div));
  });

  it('should serialize content', async () => {
    await render(
      document.body,
      <div>
        <LexicalScope />
      </div>
    );
    await pauseContainer(document.body);
    const script = getQwikJSON(document.body)!;
    expect(JSON.parse(script.textContent!)).toMatchSnapshot();
  });

  it('should serialize cyclic graphs', () => {
    useInvoke(newInvokeContext(document, div, div), () => {
      const foo = useStore({ mark: 'foo', bar: {} });
      const bar = useStore({ mark: 'bar', foo: foo });
      foo.bar = bar;
      qDiv.foo = foo;

      pauseContainer(document);

      qDiv = getProps(getContext(div));
      const foo2 = qDiv.foo;
      const bar2 = foo2.bar;
      expect(foo2.mark).toEqual('foo');
      expect(bar2.mark).toEqual('bar');
      expect(foo2.bar.foo == foo2).toBe(true);
    });
  });
});

export const LexicalScope_render = () => {
  const [a, b, c, d, e, f, g, h, state, noserialize] = useLexicalScope();
  return (
    <section>
      <p>{JSON.stringify(a)}</p>
      <p>{JSON.stringify(b)}</p>
      <p>{JSON.stringify(c)}</p>
      <p>{String(d)}</p>
      <p>{String(e)}</p>
      <p>{JSON.stringify(f)}</p>
      <p>{JSON.stringify(g)}</p>
      <p>{JSON.stringify(h)}</p>
      <p>{noserialize.text}</p>
      <button onDocumentClick$={() => state.count++}>Rerender {state.count}</button>
    </section>
  );
};

export const LexicalScope = component$(() => {
  const state = useStore({
    count: 0,
  });
  const a = 1;
  const b = 'hola';
  const c = {
    a: { thing: 12 },
    b: 'hola',
    c: 123,
    d: false,
    e: true,
    f: null,
    g: undefined,
    h: [1, 'string', false, { hola: 1 }, ['hello']],
  };
  const noserialize = noSerialize({ text: 'not included' });
  const d = undefined;
  const e = null;
  const f = [1, 2, 'hola', {}];
  const g = true;
  const h = false;
  const qrl = $(() => logDebug('qrl'));

  const thing = runtimeQrl(LexicalScope_render, [a, b, c, d, e, f, g, h, state, noserialize, qrl]);
  return <div onClickQrl={thing}></div>;
});
