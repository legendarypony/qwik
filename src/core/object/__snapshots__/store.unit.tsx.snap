// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store should serialize content 1`] = `
Object {
  "objs": Array [
    Object {},
    Object {},
    Object {
      "count": 0,
    },
    "/runtimeQRL#s0",
    1,
    "hola",
    Object {
      "a": "7",
      "b": "5",
      "c": 123,
      "d": "i",
      "e": "h",
      "f": "e",
      "g": "d",
      "h": "8",
    },
    Object {
      "thing": 12,
    },
    Array [
      "4",
      "9",
      "i",
      "a",
      "b",
    ],
    "string",
    Object {
      "hola": "4",
    },
    Array [
      "c",
    ],
    "hello",
    "",
    null,
    Array [
      "4",
      2,
      "5",
      "g",
    ],
    Object {},
    true,
    false,
    "/runtimeQRL#s1",
  ],
  "subs": Array [
    Object {},
    Object {},
    Object {},
  ],
}
`;
